roomserver/internal/input/input.go

```diff
@@ -74,18 +105,51 @@ func (r *Inputer) InputRoomEvents(
 	request *api.InputRoomEventsRequest,
 	response *api.InputRoomEventsResponse,
 ) (err error) {
+	// Create a wait group. Each task that we dispatch will call Done on
+	// this wait group so that we know when all of our events have been
+	// processed.
+	wg := &sync.WaitGroup{}
+	wg.Add(len(request.InputRoomEvents))
+	tasks := make([]*inputTask, len(request.InputRoomEvents))
+
 	for i, e := range request.InputRoomEvents {
+		// Work out if we are running per-room workers or if we're just doing
+		// it on a global basis (e.g. SQLite).
 		roomID := "global"
 		if r.DB.SupportsConcurrentRoomInputs() {
 			roomID = e.Event.RoomID()
 		}
-		mutex, _ := r.mutexes.LoadOrStore(roomID, &sync.Mutex{})
-		mutex.(*sync.Mutex).Lock()
-		if response.EventID, err = r.processRoomEvent(ctx, request.InputRoomEvents[i]); err != nil {
-			mutex.(*sync.Mutex).Unlock()
+
+		// Look up the worker, or create it if it doesn't exist.
+		w, _ := r.workers.LoadOrStore(roomID, &inputWorker{
+			r:     r,
+			input: make(chan *inputTask),
+		})
+		worker := w.(*inputWorker)
+
+		// Create a task. This contains the input event and a reference to
+		// the wait group, so that the worker can notify us when this specific
+		// task has been finished.
+		tasks[i] = &inputTask{
+			ctx:   ctx,
+			event: e,
+			wg:    wg,
+		}
+
+		// Send the task to the worker.
+		go func(task *inputTask) { worker.input <- task }(tasks[i])
```

Comment:
You lose input ordering by doing this in a goroutine. There's no guarantee that the goroutine will pass this task into `worker.input` before the next loop